wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -149,12 +147,30 @@ protected Builder() {
      * this message has no unknown fields.
      */
     public final ByteString buildUnknownFields() {
-      return unknownFieldsBuffer != null
-          ? unknownFieldsBuffer.clone().readByteString()
-          : ByteString.EMPTY;
+      if (unknownFieldsBuffer != null) {
+        // Reads and caches the unknown fields from the buffer.
+        unknownFieldsByteString = unknownFieldsBuffer.readByteString();
+        unknownFieldsBuffer = null;
```

Comment:
I think my preference in that case would be to have `unknownFieldsByteString` be the sole source of truth, and when an unknown field is added, create a new `Buffer`, write `unknownFieldsByteString` into it, and write the results back into `unknownFieldsByteString`. Although it's slower, the intent will be clearer since there are no cache fields needed.